Add SemIR Vtable instruction and usage

dwblaikie opened the pull request:
Add a Vtable typed inst with a type_id (of the type this vtable applies to) and list of virtual function decls (or import refs to function object constants).

This doesn't add lowering/emission of the vtable, or usage when initializing objects of the type.

Some questions in case they're interesting to discuss:
* is it right/worth having the type_id in the vtable? (probably makes it easier to emit - using the type to get the class name to figure out the mangled name for the vtable) perhaps it should be a ClassId?
* I'm thinking the logic in CheckCompleteClassType could be the place we handle diagnostics for mismatched keywords (virtual/abstract for a function that's already virtual/abstract, maybe checking for non-virtual functions with the same name in a base class, or derived class functions without `impl`, etc) - but we could move some of that to the moment we walk the function decl, and record our findings in the function decl (record the base function it overrides, or the index of the vtable to slot to use when building the vtable at the end of the class)
* the Vtable typed inst has `constant_kind = InstConstantKind::Always` and `is_lowered = false`, I think I added that in to workaround/address some failures in lowering. And seems correct for this intermediate step - I'll add lowering in a follow-up patch. But the constant_kind - what should this be? We can just say all vtables are of VtableType (in which case the `Always` constant kind sounds right to me)  or we could have them introduce a type with each virtual function as a named member, even?

dwblaikie committed:
WIP SemIR vtables

danakj commented on toolchain/check/import_ref.cpp:
```diff
@@ -1392,6 +1392,26 @@ static auto TryResolveTypedInst(ImportRefResolver& resolver,
+          {.type_id =
+               resolver.local_context().GetTypeIdForTypeConstant(type_const_id),
+           .virtual_functions_id = virtual_functions_id}));
+  return ResolveResult::Done(resolver.local_constant_values().Get(inst_id));
```
`inst_id` comes from `AddInstInNoBlock` so looking it up in `local_constant_values()` looks wrong to me. Should it be `local_insts()`?

dwblaikie replied:
I tried something like this, but didn't have any luck changing the small test case outcome (& autoupdate still crashes, which may or may not be related/the same crash - mostly ignoring that until I can get a small import-and-derive test case to work)

```
diff --git a/toolchain/check/import_ref.cpp b/toolchain/check/import_ref.cpp
index 0a89777bf..b23c7ce81 100644
--- a/toolchain/check/import_ref.cpp
+++ b/toolchain/check/import_ref.cpp
@@ -1393,7 +1393,8 @@ static auto TryResolveTypedInst(ImportRefResolver& resolver,
 }

 static auto TryResolveTypedInst(ImportRefResolver& resolver, SemIR::Vtable inst,
-                                SemIR::InstId import_inst_id) -> ResolveResult {
+                                SemIR::InstId /*import_inst_id*/)
+    -> ResolveResult {
   auto type_const_id = GetLocalConstantId(resolver, inst.type_id);
   auto virtual_functions =
       GetLocalInstBlockContents(resolver, inst.virtual_functions_id);
@@ -1403,13 +1404,10 @@ static auto TryResolveTypedInst(ImportRefResolver& resolver, SemIR::Vtable inst,

   auto virtual_functions_id = GetLocalCanonicalInstBlockId(
       resolver, inst.virtual_functions_id, virtual_functions);
-  auto inst_id = resolver.local_context().AddInstInNoBlock(
-      resolver.local_context().MakeImportedLocAndInst<SemIR::Vtable>(
-          AddImportIRInst(resolver, import_inst_id),
-          {.type_id =
-               resolver.local_context().GetTypeIdForTypeConstant(type_const_id),
-           .virtual_functions_id = virtual_functions_id}));
-  return ResolveResult::Done(resolver.local_constant_values().Get(inst_id));
+  return ResolveAs<SemIR::Vtable>(
+      resolver, {.type_id = resolver.local_context().GetTypeIdForTypeConstant(
+                     type_const_id),
+                 .virtual_functions_id = virtual_functions_id});
 }
```
but, yeah, throwing things in the dark here a bit :/

zygoloid commented on toolchain/check/import_ref.cpp:
```diff
@@ -1392,6 +1392,26 @@ static auto TryResolveTypedInst(ImportRefResolver& resolver,
+
+  auto virtual_functions_id = GetLocalCanonicalInstBlockId(
+      resolver, inst.virtual_functions_id, virtual_functions);
+  auto inst_id = resolver.local_context().AddInstInNoBlock(
+      resolver.local_context().MakeImportedLocAndInst<SemIR::Vtable>(
+          AddImportIRInst(resolver, import_inst_id),
+          {.type_id =
+               resolver.local_context().GetTypeIdForTypeConstant(type_const_id),
+           .virtual_functions_id = virtual_functions_id}));
+  return ResolveResult::Done(resolver.local_constant_values().Get(inst_id));
```
We generally only add a new instruction when importing declaration-like things where the declaration and its constant value might be separate, it might be found by name lookup, it might be redeclared, that sort of thing. If you only want to import a constant value for the vtable (which I think is what we want here), you can directly produce a constant value instead of adding an instruction and then getting its constant value:
```suggestion
  return ResolveAs<SemIR::Vtable>(
      resolver,
      {.type_id =
           resolver.local_context().GetTypeIdForTypeConstant(type_const_id),
       .virtual_functions_id = virtual_functions_id}));
```

dwblaikie replied:
Ah, right - I guess I tried something like this to start, encountered problems, then went down the rabbit hole of ResolveResult, etc, trying to find something that works.

So stepping back (based on the suggested edit you provided) to this simpler implementation, I should describe the problem I encounter (oh, right, I did describe it over here: https://github.com/carbon-language/carbon-lang/discussions/4671#discussioncomment-11634477 - sorry about the bifurcated discussion, repeating that here):

```
package Modifiers;
base class B1 { virtual fn H(); }
```
```
import Modifiers;
class D1 { extend base: Modifiers.B1; }
base class B2 { virtual fn H(); }
class D2 { extend base: B2; }
```
where D2 has the correct vtable:
```
%H.decl: %H.type.2 = fn_decl @H.2 [template = constants.%H.2] {} {}
 ...
%.loc4_29: <vtable> = vtable (@B2.%H.decl) [template = constants.%.2]
```
but D1 has this weird vtable:
```
 %H.1: %H.type.1 = struct_value () [template]
 ...
 %.loc2_39: <vtable> = vtable (constants.%H.1) [template = constants.%.1]
```

Which then lead to a CHECK-fail in the commented out vtable initialization code in this PR, here:
```
       auto base_vtable_inst_block = context.inst_blocks().Get(
           context.insts()
               .GetAs<SemIR::Vtable>(base_class_info->vtable_id)
               .virtual_functions_id);
       for (auto fn_decl_id : base_vtable_inst_block) {
───────> auto fn_decl = context.insts().GetAs<SemIR::FunctionDecl>(fn_decl_id);
```
```
CHECK failure at ./toolchain/sem_ir/inst.h:182: Is<TypedInst>(): Casting inst {kind: StructValue, arg0: inst_block_empty, type: type(inst22)} to wrong kind FunctionDecl
```

dwblaikie committed:
Break cycle in import of vtable_id using ImportRef

dwblaikie committed:
Test updates

dwblaikie committed:
Handle function overrides, including across module boundaries

dwblaikie committed:
Remove vtable_id from sem_ir dump

Like the base_id, it isn't rendered separately - assumed that the base
decl in the class inst list is the base, without needing to dump that
explicitly.

dwblaikie committed:
More clear testing

dwblaikie committed:
Update tests

zygoloid commented on toolchain/check/handle_class.cpp:
```diff
@@ -696,6 +700,46 @@ static auto CheckCompleteClassType(Context& context, Parse::NodeId node_id,
+    llvm::SmallVector<SemIR::InstId> vtable;
+    if (!defining_vptr) {
+      LoadImportRef(context, base_class_info->vtable_id);
+      auto vtable_id = context.constant_values().GetConstantInstId(
+          base_class_info->vtable_id);
+      auto base_vtable_inst_block = context.inst_blocks().Get(
+          context.insts().GetAs<SemIR::Vtable>(vtable_id).virtual_functions_id);
```
```suggestion
      auto base_vtable_id = context.constant_values().GetConstantInstId(
          base_class_info->vtable_id);
      auto base_vtable_inst_block = context.inst_blocks().Get(
          context.insts().GetAs<SemIR::Vtable>(base_vtable_id).virtual_functions_id);
```
maybe?

zygoloid commented on toolchain/check/handle_class.cpp:
```diff
@@ -696,6 +700,46 @@ static auto CheckCompleteClassType(Context& context, Parse::NodeId node_id,
+          context.insts().GetAs<SemIR::Vtable>(vtable_id).virtual_functions_id);
+      for (auto fn_decl_id : base_vtable_inst_block) {
+        auto fn_decl = GetCalleeFunction(context.sem_ir(), fn_decl_id);
+        auto fn = context.functions().Get(fn_decl.function_id);
```
```suggestion
        const auto& fn = context.functions().Get(fn_decl.function_id);
```
Avoid copying the `Function` object.

zygoloid commented on toolchain/check/handle_class.cpp:
```diff
@@ -696,6 +700,46 @@ static auto CheckCompleteClassType(Context& context, Parse::NodeId node_id,
+      for (auto fn_decl_id : base_vtable_inst_block) {
+        auto fn_decl = GetCalleeFunction(context.sem_ir(), fn_decl_id);
+        auto fn = context.functions().Get(fn_decl.function_id);
+        for (auto override_fn_decl_id :
+             context.vtable_stack().PeekCurrentBlockContents()) {
+          auto override_fn_decl =
+              context.insts().GetAs<SemIR::FunctionDecl>(override_fn_decl_id);
+          auto override_fn =
+              context.functions().Get(override_fn_decl.function_id);
+          if (override_fn.virtual_modifier ==
+                  SemIR::FunctionFields::VirtualModifier::Impl &&
+              override_fn.name_id == fn.name_id) {
+            fn_decl_id = override_fn_decl_id;
+          }
+        }
```
Please add a TODO to avoid the quadratic time search here. (Eg, we could build a map from `NameId` to `InstId` first.)

dwblaikie replied:
63c943575200906a0d757e56161b7fc51b5f42c6

dwblaikie replied:
63c943575200906a0d757e56161b7fc51b5f42c6

zygoloid commented on toolchain/check/handle_class.cpp:
```diff
@@ -696,6 +700,46 @@ static auto CheckCompleteClassType(Context& context, Parse::NodeId node_id,
+            fn_decl_id = override_fn_decl_id;
+          }
+        }
+        vtable.push_back(fn_decl_id);
```
Please add a TODO to check the signatures match.

dwblaikie replied:
Ah, yep - added that in 63c943575200906a0d757e56161b7fc51b5f42c6 - actually looking forward to/was planning to add that in follow-up.

zygoloid commented on toolchain/check/handle_class.cpp:
```diff
@@ -696,6 +700,46 @@ static auto CheckCompleteClassType(Context& context, Parse::NodeId node_id,
+             context.vtable_stack().PeekCurrentBlockContents()) {
+          auto override_fn_decl =
+              context.insts().GetAs<SemIR::FunctionDecl>(override_fn_decl_id);
+          auto override_fn =
```
```suggestion
          const auto& override_fn =
```

zygoloid commented on toolchain/check/handle_class.cpp:
```diff
@@ -696,6 +700,46 @@ static auto CheckCompleteClassType(Context& context, Parse::NodeId node_id,
+
+    for (auto inst_id : context.vtable_stack().PeekCurrentBlockContents()) {
+      auto fn_decl = context.insts().GetAs<SemIR::FunctionDecl>(inst_id);
+      auto fn = context.functions().Get(fn_decl.function_id);
```
```suggestion
      const auto& fn = context.functions().Get(fn_decl.function_id);
```

zygoloid commented on toolchain/check/handle_function.cpp:
```diff
@@ -209,18 +209,19 @@ static auto BuildFunctionDecl(Context& context,
           .Case(KeywordModifierSet::Impl,
                 SemIR::Function::VirtualModifier::Impl)
           .Default(SemIR::Function::VirtualModifier::None);
+  SemIR::Class* class_info = nullptr;
```
What do you think about renaming this to `virtual_class_info` or `vptr_class_info` or something, given that we only set it if this function is virtual, and below we assume that it being set means this function is virtual.

dwblaikie replied:
Yeah, that's fair, done in ae421105c174f0e05f21e693c3fb2c44293bb9b8 (& removed the extra `is_dynamic` check below, since it's only non-null when it is dynamic anyway)

zygoloid reviewed (approved):
Thanks, looks good. Only a few tiny things.

dwblaikie committed:
rename local vtable_id to base_vtable_id for clarity

Co-authored-by: Richard Smith <richard@metafoo.co.uk>

dwblaikie committed:
Avoid a copy of a `SemIR::Function`

Co-authored-by: Richard Smith <richard@metafoo.co.uk>

jonmeow commented on toolchain/lower/handle.cpp:
```diff
@@ -310,4 +310,7 @@ auto HandleInst(FunctionContext& context, SemIR::InstId inst_id,
                        llvm::PointerType::get(context.llvm_context(), 0)));
 }
 
+auto HandleInst(FunctionContext& /*context*/, SemIR::InstId /*inst_id*/,
+                SemIR::Vtable /*inst*/) -> void {}
+
```
```suggestion
```

Given the lowering settings on `Vtable`, can this be removed?

dwblaikie replied:
Ah, yep - removed 35b27ffe77b4bd2a6f4591bc57b064bd3e4920dc (oh, didn't realize you'd made the suggested edit too... or I would've used that).

jonmeow commented on toolchain/check/context.h:
```diff
@@ -774,6 +776,8 @@ class Context {
   // The stack of generic regions we are currently within.
   GenericRegionStack generic_region_stack_;
 
+  InstBlockStack vtable_stack_;
```
Can you add a comment to document this and its semantics? Like I don't know if this is right:

```suggestion
  // Contains a vtable block for each `class` scope which is currently being defined, regardless of whether the class can have virtual functions.
  InstBlockStack vtable_stack_;
```

jonmeow reviewed (commented):
Just a couple drive-by comments, I was looking out of curiosity on the approach.

dwblaikie committed:
Avoid copying a Function object

Co-authored-by: Richard Smith <richard@metafoo.co.uk>

dwblaikie committed:
Avoid copying a Function object

Co-authored-by: Richard Smith <richard@metafoo.co.uk>

dwblaikie committed:
Add some TODOs

dwblaikie committed:
Rename class_info to virtual_class_info

dwblaikie committed:
Remove unused HandleInst

dwblaikie committed:
Add comment for `vtable_stack_`

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

dwblaikie committed:
Line wrapping

dwblaikie committed:
Update test

dwblaikie committed:
Test update

dwblaikie committed:
Remove zombie test

This test was renamed in 0d70091bdab7dd945966da0691de107dbff7c7da and
resurrected in a merge to this PR.

josh11b commented on toolchain/check/import_ref.cpp:
```diff
@@ -2717,6 +2742,9 @@ static auto TryResolveInstCanonical(ImportRefResolver& resolver,
       return TryResolveTypedInst(resolver, inst);
     }
+    case CARBON_KIND(SemIR::Vtable inst): {
+      return TryResolveTypedInst(resolver, inst, inst_id);
```
Why pass `inst_id` if it isn't going to be used?

dwblaikie replied:
Ah, thanks for the catch - removed the unused parameter in #4832
